<!-- This is a test for iri_robot_pose_ekf using a bag recorded with TEO robot-->
<launch>

  <arg name="bag" default="/home/mmorta/Experiments/PAU/odom-testing-basics/rampeta-mobilelab.bag" />

  <param name="/use_sim_time" value="True"/>

  <include file="$(find teo_base)/launch/teo_state_publisher.launch" />

  <node name="rosbag" type="rosbag" pkg="rosbag" output="screen"
        args = "play $(arg bag) --clock"
        >
    <remap from="/tf" to="/tf_live"/>
  </node>


  <!-- TOPICS FROM THE BAG (among others):

    /teo/front_laser/scan
      frame_id: teo/front_laser
      sensor_msgs/LaserScan

    /teo/segway/status
      iri_segway_rmp400/SegwayRMP400Status

    /teo/tcm3/tcm3
      frame_id: teo/tcm3
      iri_common_drivers_msgs/compass3axis

  -->

  <!-- Segway -> Exmaps Odometry-->

  <node name="exmaps_odom"
        pkg="iri_segway_rmp400_odom"
        type="segway_rmp400_odom_3Dexpmaps"
        output="log"
        >
        <remap from="~/platform_status" to="/teo/segway/status"/>
        <remap from="~/odom" to="/teo/odom_exmaps"/>
        <param name="tf_prefix" value="teo"/>
        <param name="publish_tf" value="false" />
  </node>

  <!-- Front Laser -> ICP -> pose2D to Odometry -->

    <node
        name="front_icp"
        pkg="iri_laser_scan_matcher"
        type="laser_scan_matcher_node"
        output="screen"
        >
        <remap from="/scan" to="/teo/front_laser/scan"/>
        <remap from="/odom" to="/teo/odom_exmaps"/>
        <remap from="/pose2D" to="/teo/pose2D_front"/>
        <param name="fixed_frame" value="/teo/odom"/>
        <param name="base_frame"  value="/teo/base_footprint"/>
        <param name="use_imu"        value="false"/>
        <param name="use_odom"       value="true"/>
        <param name="use_alpha_beta" value="false"/>
        <param name="max_iterations" value="10"/>
        <param name="publish_tf"   value="false"/>
        <param name="publish_pose" value="false"/>
  </node>
  <node name="front_pose2d_to_odom"
        pkg="iri_pose2d_odom"
        type="iri_pose2d_odom"
        output="screen"
        >
        <remap from="~/pose2d" to="/teo/pose2D_front" />
        <remap from="~/pose2d_odom" to="/teo/odom_icp_front" />
        <param name="parent_id"  value="/teo/odom" />
        <param name="frame_id"   value="/teo/base_footprint" />
        <param name="publish_tf" value="false" />
  </node>

  <!-- Compass -> compass to Odometry -->

  <node name="compass_to_odom"
        pkg="iri_compass_odom"
        type="iri_compass_odom"
        output="screen"
        >
        <remap from="~/compass" to="/teo/tcm3/tcm3"/>
        <remap from="~/compass_odom" to="/teo/odom_compass"/>
        <param name="parent_id"  value="/teo/odom" />
  </node>

  <!-- Odometries from segway, front laser and compass -> EKF -->

  <node pkg="iri_robot_pose_ekf" type="iri_robot_pose_ekf" name="ekf_odom" respawn="true" output="screen">
    
    <param name="freq" value="30.0"/>
    <param name="sensor_timeout" value="5.0"/>
    <param name="odom_used" value="true"/>
    <param name="icp_used" value="true"/>
    <param name="cmp_used" value="true"/>
    <param name="gps_used" value="false"/>

    <remap from="/odom_data" to="/teo/odom_exmaps" />
    <remap from="/icp_data" to="/teo/odom_icp_front" />
    <remap from="/cmp_data" to="/teo/odom_compass" />
    <param name="output_frame" value="/teo/pose_combined"/>
    <!--remap from="/odom_combined" to="/teo/odom_ekf" /-->
  </node>

      <node name="pose_to_odom"
        pkg="iri_posewithcovariancestamped_odom"
        type="iri_posewithcovariancestamped_odom"
        output="screen"
        >
        <remap from="~/pose" to="/teo/pose_combined"/>
        <remap from="~/odom" to="/teo/odom"/>
        <param name="child_id"  value="/teo/odom" />
  </node>

</launch>

